zebra-state/src/service/read/tests/vectors.rs

```diff
@@ -100,6 +100,115 @@ async fn populated_read_state_responds_correctly() -> Result<()> {
+        for db_subtree in db_subtrees {
+            let mut db_batch = DiskWriteBatch::new();
+            db_batch.insert_sapling_subtree(&db, &db_subtree);
+            db.write(db_batch)
+                .expect("Writing a batch with a Sapling subtree should succeed.");
+        }
+        db
+    };
+
+    // Prepare the non-finalized state.
+    let chain = {
+        let mut chain = Chain::default();
+        let chain_subtrees = chain_height_range
+            .enumerate()
+            .map(|(index, height)| dummy_subtree((index_offset + index, height)));
+
+        for chain_subtree in chain_subtrees {
+            chain.insert_sapling_subtree(chain_subtree);
+        }
+
+        Arc::new(chain)
+    };
+
+    let modify_chain = |chain: &Arc<Chain>, index: usize, height| {
+        let mut chain = chain.as_ref().clone();
+        chain.insert_sapling_subtree(dummy_subtree((index, height)));
+        Some(Arc::new(chain))
+    };
+
+    // There should be 10 entries in db and 2 in chain with no overlap
+
+    // Unbounded range should start at 0
+    let all_subtrees = sapling_subtrees(Some(chain.clone()), &db, ..);
+    assert_eq!(all_subtrees.len(), 12, "should have 12 subtrees in state");
+
+    // Add a subtree to `chain` that overlaps and is not consistent with the db subtrees
+    let first_chain_index = index_offset - 1;
+    let end_height = Height(400_000);
+    let modified_chain = modify_chain(&chain, first_chain_index, end_height.0);
+
+    // The inconsistent entry and any later entries should be omitted
+    let all_subtrees = sapling_subtrees(modified_chain.clone(), &db, ..);
+    assert_eq!(all_subtrees.len(), 10, "should have 10 subtrees in state");
+
+    let first_chain_index =
+        NoteCommitmentSubtreeIndex(u16::try_from(first_chain_index).expect("should fit in u16"));
+
+    // Entries should be returned without reading from disk if the chain contains the first subtree index in the range
+    let mut chain_subtrees = sapling_subtrees(modified_chain, &db, first_chain_index..);
+    assert_eq!(chain_subtrees.len(), 3, "should have 3 subtrees in chain");
+
+    let (index, subtree) = chain_subtrees
+        .pop_first()
+        .expect("chain_subtrees should not be empty");
+    assert_eq!(first_chain_index, index, "subtree indexes should match");
+    assert_eq!(end_height, subtree.end, "subtree end heights should match");
+
+    // Check that Zebra retrieves subtrees correctly when using a range with an Excluded start bound
+
+    #[derive(Clone, Default, PartialEq, Eq, Hash)]
+    struct RangeExclusiveStart<Idx> {
+        pub start: Idx,
+        pub end: Option<Idx>,
+    }
+
+    impl<T> std::ops::RangeBounds<T> for RangeExclusiveStart<T> {
+        fn start_bound(&self) -> Bound<&T> {
+            Excluded(&self.start)
+        }
+        fn end_bound(&self) -> Bound<&T> {
+            self.end.as_ref().map(Excluded).unwrap_or(Unbounded)
+        }
+    }
+
+    let range = RangeExclusiveStart::default();
```

Comment:
These 15 lines of code can be replaced with one, because RangeBounds is implemented for a tuple pair:
https://doc.rust-lang.org/std/ops/trait.RangeBounds.html#impl-RangeBounds%3CT%3E-for-(Bound%3CT%3E,+Bound%3CT%3E)

Normally I'd say this simplification is optional, but it's actually tricky to read the test without it.

```rust
    let range = (Excluded(0.into()), Unbounded);
```